Translate hooks-effect (Using the Effect Hook) (#54)
reactjs/ja.reactjs.org: content/docs/hooks-effect.md

}
```

> このスニペットは [1 つ前のページのカウンターの例](/docs/hooks-state.html)に基づいていますが、新しい機能をひとつ加えてあります。ドキュメントのタイトルを、クリック回数を含んだカスタムのメッセージにセットしているのです。

Reviewer: ```suggestion
このスニペットは [1 つ前のページのカウンターの例](/docs/hooks-state.html)に基づいていますが、新しい機能をひとつ加えてあります。クリック回数を含んだカスタムのメッセージをドキュメントのタイトルにセットしているのです。
```

> データの取得、購読の設定、あるいは React コンポーネント内の DOM の手動での変更、といったものはすべて副作用の例です。これらを "副作用 (side-effect)"（あるいは単に "作用 (effect)"）と呼ぶことに慣れているかどうかはともかくとしても、このような作業自体はやったことがあるでしょう。

Reviewer: ```suggestion
データの取得、購読の設定、あるいは React コンポーネント内の DOM の手動での変更、といったものはすべて副作用の例です。これらを "副作用 (side-effect)"（あるいは単に "作用 (effect)"）と呼ぶことに慣れているかどうかはともかくとしても、これらのようなことをコンポーネントに対して行ったことはあるでしょう。
```

> ヒント
>
> React のライフサイクルに馴染みがある場合は、`useEffect` フックを `componentDidMount` と `componentDidUpdate` と `componentWillUnmount` がまとまったものだと考えることができます。

React コンポーネントにおける副作用には 2 種類あります。クリーンアップコードを必要としない副作用と、必要とする副作用です。これらの違いについて詳しく見ていきましょう。

## クリーンアップを必要としない副作用

> 時に、**React が DOM を更新した後で追加のコードを走らせたい**という場合があります。ネットワークリクエストの送信、手動での DOM 改変、ログの記録、といったものがクリーンアップを必要としない副作用の例です。なぜかというとそれらのコードを走らせたあとはすぐにそのことを忘れても構わないからです。クラスとフックとでそのような副作用をどのように表現するのか比較してみましょう。

Reviewer: ```suggestion
時に、**React が DOM を更新した後で追加のコードを実行したい**という場合があります。ネットワークリクエストの送信、手動での DOM 改変、ログの記録、といったものがクリーンアップを必要としない副作用の例です。なぜかというとそれらのコードが実行されたあとすぐにそのことを忘れても構わないからです。クラスとフックとでそのような副作用をどのように表現するのか比較してみましょう。
```

### クラスを使った例

React のクラスコンポーネントでは、`render` メソッド自体が副作用を起こすべきではありません。そこで副作用を起こすのは早すぎます -- 典型的には、副作用は React が DOM を更新した*あと*に起こすようにしたいのです。

> ですので React のクラスでは、副作用は `componentDidMount` と `componentDidUpdate` に記載します。例に戻ると、以下が React のクラスで実装したカウンターコンポーネントであり、React が DOM に変更を加えた後に、ドキュメントのタイトルを更新しています。

Reviewer: ```suggestion
そのため React のクラスでは、副作用は `componentDidMount` と `componentDidUpdate` に記載します。例に戻ると、以下が React のクラスで実装したカウンターコンポーネントであり、React が DOM に変更を加えた後に、ドキュメントのタイトルを更新しています。
```

}
```

ここで**同じコードを 2 回書かなければならなかった**ことに注意してください。

これは、コンポーネントがマウント直後なのか更新後なのかに関係なく、大抵の場合は同じ副作用を起こしたいからです。概念的には、毎回のレンダー時に起こってほしいのですが、React のクラスコンポーネントにはそのようなメソッドは存在していません。副作用のコードを別のメソッドに抽出することは可能ですが、2 か所でそのメソッドを呼ばなければいけないことに変わりはありません。

> では、同じことが `useEffect` フックを用いればどのように可能なのか見ていきましょう。

Reviewer: ```suggestion
では、同じことが `useEffect` フックを用いるとどのように記述できるのか見ていきましょう。
```

}
```

**`useEffect` は何をやっているのか？**　このフックを使うことで、レンダー後に何かの処理をしないといけない、ということを React に伝えます。React はあなたが渡した関数を覚えており（これを「副作用（関数）」と呼ぶこととします）、DOM の更新の後にそれを呼び出します。この副作用の場合はドキュメントのタイトルをセットしていますが、データを取得したりその他何らかの命令型の API を呼び出したりすることも可能です。

> **`useEffect` がコンポーネント内で呼ばれるのはなぜか？**　コンポーネント内で `useEffect` を記述することで、副作用内から state である `count`（や任意の props）にアクセスできるようになります。それらは既に関数スコープ内に存在するので、読みだすために特別な API は必要ありません。フックは JavaScript のクロージャを活用しており、JavaScript で解決できることに対して React 特有の API を導入することはしません。

Reviewer: ```suggestion
**`useEffect` がコンポーネント内で呼ばれるのはなぜか？**　コンポーネント内で `useEffect` を記述することで、副作用内から state である `count`（や任意の props）にアクセスできるようになります。それらは既に関数スコープ内に存在するので、参照するための特別な API は必要ありません。フックは JavaScript のクロージャを活用しており、JavaScript で解決できることに対して React 特有の API を導入することはしません。
```参照するための

  });
```

> ここでは `count` という state 変数を宣言し、次に副作用を使うことを React に伝えています。`useEffect` フックには関数を渡しています。この関数こそが副作用関数です。この副作用関数内で `document.title` というブラウザ API を使ってドキュメントのタイトルを設定しています。副作用関数は関数スコープ内にあるため最新の `count` の値は副作用内から参照可能です。React がコンポーネントをレンダーする際に React はこの副作用を覚えておき、DOM を更新した後に呼び出します。これは初回を含む毎回のレンダー時に発生します。

Reviewer: ```suggestion
ここでは `count` という state 変数を宣言し、次に副作用を使うことを React に伝えています。`useEffect` フックには関数を渡しています。このフックに渡された関数こそが副作用関数です。この副作用関数内で `document.title` というブラウザ API を使ってドキュメントのタイトルを設定しています。副作用関数は関数スコープ内にあるため最新の `count` の値は副作用内から参照可能です。React がコンポーネントをレンダーする際、React は実行された副作用を覚えており、DOM を更新した後にその副作用を呼び出します。これは初回を含む毎回のレンダー時に発生します。
```

Author: 「実行された副作用を覚えており」はおかしいです、Reactは useEffect が呼ばれた時点では副作用は単に（後の実行のために）その副作用関数を記憶しているだけであり実行はしていないからです

Reviewer: たしかにそのとおりでした。ご指摘ありがとうございます。
これはレビューではなく相談というか疑問というかなんですが、`it will remember the effect we used` の `we used` ってどういう意味なんでしょう…？
直訳すると「私たちが使った」ですが、この関数はまだ使われてないですよね？

Author: we used は深く考えていませんでしたが「useEffectに渡すのに使った」というだけかなと…たぶん

Reviewer: なるほど…俗語的な感じなのでしょうか…
これが謎だなと思っていました。
お答えいただきありがとうございます！

経験のある JavaScript 開発者であれば、`useEffect` に渡される関数は毎回のレンダーごとに異なっているということに気付くかもしれません。これは意図的なものです。むしろ、そのようにすることで、古い値を参照してしまう心配なしに副作用関数内から `count` を読むことができるのです。再レンダーごとに、React は*違う*副作用関数をスケジュールし、前のものを置き換えます。ある意味で、こうすることで副作用はレンダーの結果の一部のようにふるまうようになります -- それぞれの副作用は特定のひとつのレンダーと結びついているのです。これがなぜ便利なのかについては[このページの後半](#explanation-why-effects-run-on-each-update)で明らかになるでしょう。

> ヒント
>

> > `componentDidMount` や `componentDidUpdate` と異なり、`useEffect` でスケジュールされた副作用はブラウザによる画面更新をブロックしません。このためアプリの反応がよりよくなると感じられます。大部分の副作用は同期的に行われる必要がありません。同期的に行う必要がある稀なケース（レイアウトの測定など）のために、`useEffect` と同一の API を有する [`useLayoutEffect`](/docs/hooks-reference.html#uselayouteffect) という別のフックがあります。

Reviewer: ```suggestion
> `componentDidMount` や `componentDidUpdate` と異なり、`useEffect` でスケジュールされた副作用はブラウザによる画面更新をブロックしません。このためアプリの反応がより良く感じられます。大部分の副作用は同期的に行われる必要がありません。同期的に行う必要がある稀なケース（レイアウトの測定など）のために、`useEffect` と同一の API を有する [`useLayoutEffect`](/docs/hooks-reference.html#uselayouteffect) という別のフックがあります。
```

## クリーンアップを有する副作用

> ここまでに、クリーンアップを必要としない副作用の表現のしかたについて見てきました。しかし幾つかの副作用ではそれが必要です。例えば何らかの外部のデータソースへの**購読をセットアップしたい**ことがあります。そのような場合、メモリーリークを起こさないようにクリーンアップが必要です！　クラスの場合とフックの場合とでそれらをどのように行えるのか比較しましょう。

Reviewer: ```suggestion
ここまでに、クリーンアップを必要としない副作用の表現のしかたについて見てきました。しかし幾つかの副作用ではそれが必要です。例えば何らかの外部のデータソースへの**購読をセットアップしたい**ことがあります。そのような場合、メモリーリークが発生しないようにクリーンアップが必要です！ クラスの場合とフックの場合とでクリーンアップをどのように行えるのか比較しましょう。
```

}
```

ここで、`componentDidMount` と `componentWillUnmount` とがお互いに鏡のように対応していないといけないことに注意してください。ライフサイクルメソッドを使うと、2 つのメソッドに書かれているコードが概念上は同一の副作用に関連しているとしても、それらを分割して書かないといけません。

> 補足
>
> 目ざとい読者なら、この例が完全に正しいものであるためには `componentDidUpdate` も必要だと気付くかもしれません。今のところは気にしないでおきますが、このページの[後に出てくる節](#explanation-why-effects-run-on-each-update)で改めて説明します。

### フックを使った例

このコンポーネントをフックを使ってどのように書けるのか見ていきましょう。

> クリーンアップ用の別の副作用が必要だとお考えかもしれません。しかし購読を開始するコードと解除するコードとは密に関連しているため、`useEffect` はそれらを一緒に書けるようにデザインされています。あなたの副作用が関数を返した場合、React はクリーンアップのタイミングが来たらそれを実行するのです。

Reviewer: ```suggestion
クリーンアップ用の別の副作用が必要だとお考えかもしれません。しかし購読を開始するコードと解除するコードとは密に関連しているため、`useEffect` はそれらを一緒に書けるようにデザインされています。あなたの副作用が関数をした場合、React はクリーンアップのタイミングが来たらそれを実行します。
```

Author: 「関数をした」とは？

Reviewer: ごめんなさい、これは完全に僕の間違いです。

}
```

> **副作用内からなぜ関数を返したのか？**　これこそが副作用のクリーンアップのためのオプションの仕組みです。すべての副作用は、それをクリーンアップするための関数を返すことができます。これにより購読を開始するためのロジックと解除するためのロジックを並べて書くことができます。両方とも同じ副作用の一部なのです！

Reviewer: ```suggestion
**副作用内からなぜ関数を返したのか？**　これこそが副作用のクリーンアップのためのオプションの仕組みです。すべての副作用は、それをクリーンアップするための関数を返すことがあります。これにより購読を開始するためのロジックと解除するためのロジックを並べて書くことができます。両方とも同じ副作用の一部なのです！
```

Author: これは普通に「できます」でいいと思います、「クリーンアップがやりたい！」という文脈での話ですので

Reviewer: おっしゃることは理解できるのですが、 `may` を「ことができる」と翻訳することに違和感があります。
ここが `can` であればそれで問題ないのですが、 `may` には不確実性が含まれるからです。
`「それぞれの副作用はそれをクリーンアップするための関数を返す場合があります。」`
などはどうでしょう。

Author: may には「～することができる、～して（も）よい、～して差し支えない」という許可の意味があるはずです。逆に、「副作用が関数を返す場合があります」あるいは「返すことがあります」だと、自分のコードの話ではなく、読者が制御できない事象について述べているような違和感を感じます。

Reviewer: なるほどー。
許可の意味があるのは理解していますが「～することができる」というのは許可ではなく可能だと考えていたので違和感を感じていました。

> 読者が制御できない事象について述べているような違和感
言われてみればたしかにそうですね。

元の翻訳で良いと思います！
お手数おかけし申し訳ありません！

Author: ありがとうございました！

> **React は具体的には副作用のクリーンアップをいつ発生させるのか？**　React はコンポーネントがアンマウントされるときにクリーンアップを実行します。しかし、すでに学んだ通り、副作用は 1 回だけでなく毎回のレンダー時に実行されます。このため React は、ひとつ前のレンダーによる副作用を、次回の副作用を走らせる前に*も*クリーンアップします。この後で、これが[なぜバグの回避につながるのか](#explanation-why-effects-run-on-each-update)、そして[パフォーマンスの問題がある場合にどうのようにしてこの挙動を止めるのか](#tip-optimizing-performance-by-skipping-effects)について説明します。

Reviewer: ```suggestion
**React は具体的には副作用のクリーンアップをいつ発生させるのか？**　React はコンポーネントがアンマウントされるときにクリーンアップを実行します。しかし、すでに学んだ通り、副作用は 1 回だけでなく毎回のレンダー時に実行されます。このため React は、ひとつ前のレンダーによる副作用を、次回の副作用を実行する前に*も*クリーンアップします。この後で、これが[なぜバグの回避につながるのか](#explanation-why-effects-run-on-each-update)、そして[これがパフォーマンスの問題を引き起こしている場合にどうのようにしてこの挙動を止めるのか](#tip-optimizing-performance-by-skipping-effects)について説明します。
```

> 補足
>
> 副作用から名前付きの関数を返す必要はありません。ここでは目的を明示するために `cleanup` という名前にしましたが、アロー関数を返すことも別の名前を付けることも可能です。

## まとめ

> コンポーネントのレンダー後に走る様々な種類の副作用を表現するために `useEffect` が使えるということを学びました。いくつかの副作用にはクリーンアップが必要ですので、副作用は関数を返すことができます：

Reviewer: ```suggestion
`useEffect` を用いることでコンポーネントのレンダー後に実行される様々な種類の副作用を表現できることを学びました。いくつかの副作用ではクリーンアップが必要な可能性があり、クリーンアップが必要な副作用は関数を返します：
```

}
```

フックを使うことで、ライフサイクルのメソッド名に基づくのではなく、**実際に何をやっているのかに基づいてコードを分割できます**。React はコンポーネントで利用されている*すべての*副作用を、指定されている順番で適用していきます。

> ### 解説：なぜ副作用は毎回の更新ごとに走るのか

Reviewer: ```suggestion
### 解説：なぜ副作用は毎回の更新ごとに実行されるのか
```

クラスに慣れていれば、なぜクリーンアップフェーズは、アンマウント時の 1 度だけではなく再レンダー時に毎回発生するのか、と不思議に思っているかもしれません。実践的な例で、この設計によりなぜバグの少ないコンポーネントが作れるようになるのか見てみましょう。

> [このページの前の部分](#example-using-classes-1)で、フレンドがオンラインかどうかを表示する `FriendStatus` コンポーネントの例を示しました。このクラスでは `this.props` の中にある `friend.id` を読み出して、コンポーネントがマウントした後にフレンドのステータスを購読し、アンマウント時には購読を解除します：

Reviewer: ```suggestion
[このページの前の部分](#example-using-classes-1)で、フレンドがオンラインかどうかを表示する `FriendStatus` コンポーネントの例を示しました。このクラスでは `this.props` の中にある `friend.id` を参照して、コンポーネントがマウントした後にフレンドのステータスを購読し、アンマウント時には購読を解除します：
```

  }
```

> ですがコンポーネントが表示されている最中に **`friend` プロパティが変わったらどうなるのでしょう**か？　このコンポーネントは間違ったフレンドのオンラインステータスを表示し続けてしまいます。これはバグです。しかもアンマウント時に誤ったフレンド ID を使うことになるため、購読解除の呼び出しはメモリリークやクラッシュを引き起こしてしまいます。

Reviewer: ```suggestion
ですがコンポーネントが表示されている最中に **`friend` プロパティが変わったらどうなるのでしょう**か？　このコンポーネントは間違ったフレンドのオンラインステータスを表示し続けてしまいます。これはバグです。しかも誤ったフレンド ID を使って購読解除を呼び出してしまうため、アンマウント時にメモリリークやクラッシュを引き起こしてしまうでしょう。
```

  }
```

適切な `componentDidUpdate` 処理をし忘れることが、React アプリケーションにおけるよくあるバグの原因となっていました。

> ではフック利用バージョンのこのコンポーネントを見てみましょう。

Reviewer: ```suggestion
ではこのコンポーネントのフックを利用したバージョンを見てみましょう。
```

  });
```

> 動作は変わっておらず、前述のバグも起こらなくなります。

Reviewer: ```suggestion
前述のバグの影響を受けることはありませんが、動作の変更も全くありません。
```

Author: ここは「が」を使うと混乱しそうなので敢えてこうしてあります。順番が重要なら「前述のバグの影響を受けることはありません**し**、動作の変更も全くありません。」で如何でしょうか。

Reviewer: なぜ「が」を使うと混乱を招くのでしょう？

Author: 「バグの影響を受けることはありませんが、…」とすると、そのあとには何か良くないことが続きそうに聞こえますが、ここでの「動作が変わってない」は、何も悪いことではないので…。

Reviewer: なるほど！
たしかにそうかもです。
`前述のバグの影響を受けることはありません**し**、動作の変更も全くありません。` で問題ないと思います！

> `useEffect` は*デフォルトで*更新を処理するため、更新のための特別なコードは不要です。新しい副作用を適用する前に、ひとつ前の副作用をクリーンアップします。これを例示するため、このコンポーネントが発生させる可能性のある購読登録と購読解除のシーケンスをお示しします：

Reviewer: ```suggestion
`useEffect` は*デフォルトで*更新を処理するため、更新のための特別なコードは不要です。新しい副作用を適用する前に、ひとつ前の副作用をクリーンアップします。これを例示するため、このコンポーネントがやがて発生させる可能性のある購読登録と購読解除のシーケンスを示します：
```

Author: この over time は「やがて」「そのうち」ではなく「経時的に」「時間軸に沿って」「時間の経過につれて」辺りだと思いますが、どれもしっくりこないのでなんとなく外した覚えがあります。「経時的に」辺りを付け加えたいと思います

> `useEffect` は*デフォルトで*更新を処理するため、更新のための特別なコードは不要です。新しい副作用を適用する前に、ひとつ前の副作用をクリーンアップします。これを例示するため、このコンポーネントが発生させる可能性のある購読登録と購読解除のシーケンスをお示しします：

Reviewer 2: ```suggestion
`useEffect` は*デフォルトで*更新を処理するため、更新のための特別なコードは不要です。新しい副作用を適用する前に、ひとつ前の副作用をクリーンアップします。これを例示するため、このコンポーネントが経時的に発生させる可能性のある購読登録と購読解除のシーケンスを示しします：
```

ChatAPI.unsubscribeFromFriendStatus(300, handleStatusChange); // Clean up last effect
```

この挙動によりデフォルトで一貫性を保証することができ、クラスコンポーネントでよく見られた更新ロジック書き忘れによるバグを防止することができます。

### ヒント：副作用のスキップによるパフォーマンス改善

> いくつかの場合、副作用のクリーンアップと適用とをレンダーごとに毎回行うことはパフォーマンスの問題を引き起こします。クラスコンポーネントの場合、この問題は `componentDidUpdate` の内部で `prevProps` や `prevState` と比較するコードを加えることで解決できました。

Reviewer: ```suggestion
いくつかの場合では、副作用のクリーンアップと適用とをレンダーごとに毎回行うことはパフォーマンスの問題を引き起こす可能性があります。クラスコンポーネントの場合、この問題は `componentDidUpdate` の内部で `prevProps` や `prevState` と比較するコードを加えることで解決できました。
```

}, [props.friend.id]); // Only re-subscribe if props.friend.id changes
```

将来的には、ビルド時の変換で第 2 引数を自動で加えられるようになるかもしれません。

> 補足
>

> > この最適化を使う場合、**副作用の内部で使われ、かつ副作用外部のスコープで変化する値**を利用するようにしてください。さもないとあなたのコードは以前のレンダー時の古い値を参照してしまうことになります。その他の最適化のオプションについては [Hooks API リファレンス](/docs/hooks-reference.html)で説明しています。

Reviewer: ```suggestion
> この最適化を使う場合、**時間の経過とともに変化し副作用によって使用される外部スコープからの値**が配列に含まれていることを確認してください。さもないとあなたのコードは以前のレンダー時の古い値を参照してしまうことになります。その他の最適化のオプションについては [Hooks API リファレンス](/docs/hooks-reference.html)で説明しています。
```
